spring-cloud-aws-sqs/src/main/java/io/awspring/cloud/sqs/listener/errorhandler/ExponentialBackoffErrorHandler.java

```diff
@@ -0,0 +1,167 @@
+ * whenever an exception occurs.
+ *
+ * <p>
+ * When AcknowledgementMode is set to ON_SUCCESS (the default), returning a failed future prevents the message from
+ * being acknowledged.
+ *
+ * @author Bruno Garcia
+ * @author Rafael Pavarini
+ */
+
+public class ExponentialBackoffErrorHandler<T> implements AsyncErrorHandler<T> {
+	private static final Logger logger = LoggerFactory.getLogger(ExponentialBackoffErrorHandler.class);
+
+	private final int initialVisibilityTimeoutSeconds;
+	private final double multiplier;
+	private final int maxVisibilityTimeoutSeconds;
+
+	private ExponentialBackoffErrorHandler(int initialVisibilityTimeoutSeconds, double multiplier,
+			int maxVisibilityTimeoutSeconds) {
+		this.initialVisibilityTimeoutSeconds = initialVisibilityTimeoutSeconds;
+		this.multiplier = multiplier;
+		this.maxVisibilityTimeoutSeconds = maxVisibilityTimeoutSeconds;
+	}
+
+	@Override
+	public CompletableFuture<Void> handle(Message<T> message, Throwable t) {
+		return applyExponentialBackoffVisibilityTimeout(message)
+				.thenCompose(theVoid -> CompletableFuture.failedFuture(t));
+	}
+
+	@Override
+	public CompletableFuture<Void> handle(Collection<Message<T>> messages, Throwable t) {
+		return applyExponentialBackoffVisibilityTimeout(messages)
+				.thenCompose(theVoid -> CompletableFuture.failedFuture(t));
+	}
+
+	private CompletableFuture<Void> applyExponentialBackoffVisibilityTimeout(Collection<Message<T>> messages) {
+		CompletableFuture<?>[] futures = ErrorHandlerVisibilityHelper.groupMessagesByReceiveMessageCount(messages)
+				.entrySet().stream().map(entry -> {
+					int timeout = calculateTimeout(entry.getKey());
+					return applyBatchVisibilityChange(entry.getValue(), timeout);
+				}).toArray(CompletableFuture[]::new);
+
+		return CompletableFuture.allOf(futures);
+	}
+
+	private CompletableFuture<Void> applyBatchVisibilityChange(Collection<Message<T>> messages, int timeout) {
+		logger.debug("Changing batch visibility timeout to {} - Messages Id {}", timeout,
+				MessageHeaderUtils.getId(messages));
+		BatchVisibility visibility = ErrorHandlerVisibilityHelper.getVisibility(messages);
+		return CompletableFutures.exceptionallyCompose(visibility.changeToAsync(timeout), throwable -> {
+			logger.warn("Failed to change batch visibility timeout to {} - Messages Id {}", timeout,
+					MessageHeaderUtils.getId(messages), throwable);
+			return CompletableFuture.failedFuture(throwable);
+		});
+	}
+
+	private CompletableFuture<Void> applyExponentialBackoffVisibilityTimeout(Message<T> message) {
+		int timeout = calculateTimeout(message);
+		Visibility visibility = ErrorHandlerVisibilityHelper.getVisibility(message);
+		logger.debug("Changing visibility timeout to {} - Message Id {}", timeout, message.getHeaders().getId());
+		return CompletableFutures.exceptionallyCompose(visibility.changeToAsync(timeout), throwable -> {
```

Comment:
The CompletableFutures class was necessary when we were using Java 8.

On Java 17 we can use the equivalent `CompletableFuture` methods directly.